src/library/filescanner.c

```diff
@@ -1600,6 +1614,302 @@ scan_metadata(const char *path, struct media_file_info *mfi)
+  if (!path)
+    {
+      DPRINTF(E_LOG, L_SCAN, "Unsupported virtual path '%s'\n", vp_playlist);
+      return LIBRARY_PATH_INVALID;
+    }
+
+  if (!check_path_in_directories(path))
+    {
+      DPRINTF(E_LOG, L_SCAN, "Path '%s' is not a virtual path for a configured (local) library directory.\n", vp_playlist);
+      return LIBRARY_ERROR;
+    }
+
+  pli = db_pl_fetch_byvirtualpath(vp_playlist);
+  if (pli && pli->type != PL_PLAIN)
+    {
+      DPRINTF(E_LOG, L_SCAN, "Playlist with virtual path '%s' already exists and is not a plain playlist.\n", vp_playlist);
+      free_pli(pli, 0);
+      return LIBRARY_ERROR;
+    }
+  free_pli(pli, 0);
+
+  pl_path = safe_asprintf("%s.m3u", path);
+  *fp = fopen(pl_path, "a");
+  if (!(*fp))
+    {
+      DPRINTF(E_LOG, L_SCAN, "Error opening file '%s' for writing\n", pl_path);
+      return LIBRARY_ERROR;
+    }
+
+  *pl_id = db_pl_id_bypath(pl_path);
+  if (*pl_id < 0)
+    {
+      dir_id = get_parent_dir_id(pl_path);
+      filename = filename_from_path(pl_path);
+      title = strip_extension(filename);
+      *pl_id = library_add_playlist_info(pl_path, title, vp_playlist, PL_PLAIN, 0, dir_id);
+      free(title);
+    }
+
+  return LIBRARY_OK;
+}
+
+static int
+playlist_add_path(FILE *fp, int pl_id, const char *path)
+{
+  int ret;
+
+  ret = fprintf(fp, "%s\n", path);
+  if (ret >= 0)
+    {
+      ret = db_pl_add_item_bypath(pl_id, path);
+    }
+
+  if (ret < 0)
+    {
+      DPRINTF(E_LOG, L_SCAN, "Failed to add path '%s' to playlist (id = %d)\n", path, pl_id);
+      return -1;
+    }
+
+  return 0;
+}
+
+static int
+playlist_add_files(FILE *fp, int pl_id, const char *virtual_path)
+{
+  struct query_params qp;
+  struct db_media_file_info dbmfi;
+  uint32_t data_kind;
+  int ret;
+
+  memset(&qp, 0, sizeof(struct query_params));
+  qp.type = Q_ITEMS;
+  qp.sort = S_ARTIST;
+  qp.idx_type = I_NONE;
+  qp.filter = sqlite3_mprintf("(f.virtual_path = %Q OR f.virtual_path LIKE '%q/%%')", virtual_path, virtual_path);
+
+  ret = db_query_start(&qp);
+  if (ret < 0)
+    {
+      db_query_end(&qp);
+      return -1;
+    }
+
+  while (((ret = db_query_fetch_file(&qp, &dbmfi)) == 0) && (dbmfi.id))
+    {
+      if ((safe_atou32(dbmfi.data_kind, &data_kind) < 0)
+	  || (data_kind == DATA_KIND_PIPE))
+	{
+	  DPRINTF(E_WARN, L_SCAN, "Item '%s' not added to playlist (id = %d), unsupported data kind\n", dbmfi.path, pl_id);
+	  continue;
+	}
+
+      ret = playlist_add_path(fp, pl_id, dbmfi.path);
+      if (ret < 0)
+	break;
+
+      DPRINTF(E_DBG, L_SCAN, "Item '%s' added to playlist (id = %d)\n", dbmfi.path, pl_id);
+    }
+
+  db_query_end(&qp);
+
+  return ret;
+}
+
+static int
+playlist_add(const char *vp_playlist, const char *vp_item)
+{
+  FILE *fp = NULL;
+  int pl_id;
+  int ret;
+
+  ret = playlist_open(&fp, &pl_id, vp_playlist);
+  if (ret < 0)
+    return ret;
+
+  if (!fp)
+    return LIBRARY_ERROR;
+
+  ret = playlist_add_files(fp, pl_id, vp_item);
+  fclose(fp);
```

Comment:
It might be nicer if fopen() and fclose() were in the same function? So that playlist_open() was instead used to check stuff and prepare a path for fopen().